File: csharp/src/Toolnexus/McpServe.cs
using System.Text.Json;
using Microsoft.Extensions.Logging.Abstractions;
using ModelContextProtocol;
using ModelContextProtocol.Protocol;
using ModelContextProtocol.Server;

namespace Toolnexus;

// ---------------------------------------------------------------------------
// MCP — INBOUND: serve a toolkit as an MCP server (SPEC §7C).
//
// The inbound mirror of the A2A `serve` (§7B). Where A2A advertises the toolkit's
// SKILLS and fulfils a Task through the whole client loop, MCP advertises the
// toolkit's UNIFIED TOOLS (every source — mcp · skill · native · http · builtin ·
// a2a) and dispatches each `tools/call` straight to <see cref="ITool.ExecuteAsync"/>.
// The calling MCP client *is* the LLM host, so there is no client, no Task, and no
// TaskStore — just `tools/list` + `tools/call` over the toolkit registry.
//
// Transport: streamable-HTTP, built on the same official <c>ModelContextProtocol</c>
// SDK the client side already uses — mounted at POST /mcp on the
// <see cref="Toolkit.ServeAsync"/> server, alongside the A2A routes (see A2AServer.cs).
//
// Mirrors the JS reference (js/src/mcpserve.ts). See ../../SPEC.md §7C and
// openspec/changes/add-mcp-inbound.
// ---------------------------------------------------------------------------

/// <summary>Opt-in MCP serve profile — the toolkit is exposed as an MCP server.</summary>
public sealed class MCPServeConfig
{
    /// <summary>Advertised server name (initialize <c>serverInfo.name</c>). Default "toolnexus".</summary>
    public string? Name { get; set; }

    /// <summary>Advertised server version. Default "0.1.0".</summary>
    public string? Version { get; set; }

    /// <summary>Subset of tool names to expose; omit ⇒ all. Unknown names are ignored.</summary>
    public IReadOnlyList<string>? Tools { get; set; }
}

/// <summary>Surfaced on each inbound <c>tools/call</c>.</summary>
public sealed record OnCallEvent(string Name, string Source, long Ms, bool IsError);

/// <summary>Callback fired per inbound <c>tools/call</c>. Host callback errors are isolated.</summary>
public delegate Task OnCall(OnCallEvent ev);

/// <summary>
/// Builds the toolkit's tools as an MCP server. The streamable-HTTP path
/// (<c>/mcp</c>) is driven by <see cref="A2AServer"/>.
/// </summary>
public static class McpServe
{
    private const string DefaultName = "toolnexus";
    private const string DefaultVersion = "0.1.0";

    /// <summary>
    /// The tools this profile exposes: every toolkit tool, filtered to <c>cfg.Tools</c>
    /// when set (unknown names in the filter are ignored, never an error).
    /// </summary>
    public static IReadOnlyList<ITool> ExposedMcpTools(IReadOnlyList<ITool> tools, MCPServeConfig? cfg)
    {
        if (cfg?.Tools == null) return tools;
        var wanted = new HashSet<string>(cfg.Tools);
        return tools.Where(t => wanted.Contains(t.Name)).ToList();
    }

    /// <summary>
    /// Build a low-level MCP <see cref="McpServer"/> over <paramref name="transport"/> exposing
    /// <paramref name="tools"/> as MCP tools. <c>tools/list</c> maps each tool's name (verbatim —
    /// already sanitized at registration), description, and inputSchema; <c>tools/call</c>
    /// dispatches to <see cref="ITool.ExecuteAsync"/> and maps the <see cref="ToolResult"/> to a
    /// <see cref="CallToolResult"/> (<c>Output</c> → text, <c>IsError</c> propagates).
    /// </summary>
    public static McpServer BuildMcpServer(
        ITransport transport, IReadOnlyList<ITool> tools, MCPServeConfig? cfg = null, OnCall? onCall = null)
        => McpServer.Create(transport, BuildOptions(tools, cfg, onCall), NullLoggerFactory.Instance, null);

    /// <summary>Assemble the <see cref="McpServerOptions"/> (serverInfo + tools capability + handlers).</summary>
    internal static McpServerOptions BuildOptions(IReadOnlyList<ITool> tools, MCPServeConfig? cfg, OnCall? onCall)
    {
        // First-name-wins index for dispatch (toolkit lists are already deduped).
        var byName = new Dictionary<string, ITool>();
        foreach (var t in tools)
            if (!byName.ContainsKey(t.Name)) byName[t.Name] = t;

        var options = new McpServerOptions
        {
            ServerInfo = new Implementation
            {
                Name = cfg?.Name ?? DefaultName,
                Version = cfg?.Version ?? DefaultVersion,
            },
            Capabilities = new ServerCapabilities { Tools = new ToolsCapability() },
        };

        options.Handlers.ListToolsHandler = (_, _) =>
            new ValueTask<ListToolsResult>(new ListToolsResult { Tools = tools.Select(ToMcpTool).ToList() });

        options.Handlers.CallToolHandler = async (ctx, ct) =>
        {
            var name = ctx.Params?.Name ?? "";
            if (!byName.TryGetValue(name, out var tool))
                throw new McpException($"Unknown tool: {name}");

            var started = System.Diagnostics.Stopwatch.GetTimestamp();
            // The call is a single tool invocation honoring the MCP request's cancellation.
            // An ExecuteAsync throw becomes an isError result — never a crash.
            ToolResult result;
            try
            {
                var args = ToArgs(ctx.Params?.Arguments);
                result = await tool.ExecuteAsync(args, new ToolContext(cancellationToken: ct)).ConfigureAwait(false);
            }
            catch (Exception e)
            {
                result = new ToolResult(e.Message ?? e.ToString(), true);
            }

            if (onCall != null)
            {
                var ms = (long)System.Diagnostics.Stopwatch.GetElapsedTime(started).TotalMilliseconds;
                try { await onCall(new OnCallEvent(tool.Name, tool.Source, ms, result.IsError)).ConfigureAwait(false); }
                catch { /* Host callback errors are isolated. */ }
            }

            return new CallToolResult
            {
                Content = new List<ContentBlock> { new TextContentBlock { Text = result.Output } },
                IsError = result.IsError,
            };
        };

        return options;
    }

    // -----------------------------------------------------------------------
    // Mapping helpers.
    // -----------------------------------------------------------------------

    private static Tool ToMcpTool(ITool t) => new()
    {
        Name = t.Name,
        Description = t.Description,
        InputSchema = SchemaElement(t.InputSchema),
    };

    /// <summary>Convert a tool's plain-dictionary JSON Schema into the SDK's <see cref="JsonElement"/> form.</summary>
    private static JsonElement SchemaElement(IDictionary<string, object?> schema)
    {
        using var doc = JsonDocument.Parse(Json.Stringify(schema));
        return doc.RootElement.Clone();
    }

    /// <summary>Convert MCP call arguments (<see cref="JsonElement"/> values) into the plain CLR model.</summary>
    private static IDictionary<string, object?> ToArgs(IDictionary<string, JsonElement>? arguments)
    {
        var args = new Dictionary<string, object?>();
        if (arguments == null) return args;
        foreach (var (k, v) in arguments) args[k] = Json.FromElement(v);
        return args;
    }
}

File: python/src/toolnexus/mcp_serve.py
"""MCP — INBOUND: serve a toolkit as an MCP server (SPEC §7C).

The inbound mirror of the A2A ``serve`` (§7B). Where A2A advertises the toolkit's
**skills** and fulfils a Task through the whole client loop, MCP advertises the
toolkit's **unified tools** (every source — mcp · skill · native · http · builtin
· a2a) and dispatches each ``tools/call`` straight to ``Tool.execute``. The calling
MCP client *is* the LLM host, so there is **no client, no Task, and no TaskStore** —
just ``tools/list`` + ``tools/call`` over the toolkit registry.

The server is exposed over **streamable-HTTP**, mounted at ``POST /mcp`` on the
``serve(addr, …)`` server, alongside the A2A routes (see serve.py). Built on the
same ``mcp`` SDK the client side already uses.

Mirrors the JS reference (``js/src/mcpserve.ts``).
"""
from __future__ import annotations

import asyncio
import inspect
from typing import Any, Awaitable, Callable, Optional, Union

import anyio
from mcp.server.lowlevel import Server
from mcp.server.streamable_http import StreamableHTTPServerTransport
from mcp.shared.exceptions import McpError
from mcp.types import INVALID_PARAMS, CallToolResult, ErrorData, TextContent
from mcp.types import Tool as McpToolDef

from .types import Tool, ToolResult

DEFAULT_NAME = "toolnexus"
DEFAULT_VERSION = "0.1.0"

# Opt-in MCP serve profile — the toolkit is exposed as an MCP server. A plain dict,
# mirroring the JS `MCPServeConfig` interface (and the wire config block):
#   { "name"?, "version"?, "tools"?: [str] }
#   name default "toolnexus", version default "0.1.0"; tools omit ⇒ all, unknown
#   names in the filter are ignored (never an error).
MCPServeConfig = dict[str, Any]

# on_call(ev) — fires per inbound tools/call. ev = { "name", "source", "ms",
# "is_error" } (snake_case, matching the port's other event dicts).
OnCall = Callable[[dict[str, Any]], Union[None, Awaitable[None]]]


def _coalesce(value: Any, default: Any) -> Any:
    """JS ``??`` semantics — the default applies only for None (not "" / falsy)."""
    return value if value is not None else default


def exposed_mcp_tools(
    tools: list[Tool], cfg: Optional[MCPServeConfig] = None
) -> list[Tool]:
    """The tools this profile exposes: every toolkit tool, filtered to ``cfg["tools"]``
    when set (unknown names in the filter are ignored, never an error).
    """
    wanted = cfg.get("tools") if cfg else None
    if wanted is None:
        return tools
    wanted_set = set(wanted)
    return [t for t in tools if t.name in wanted_set]


def build_mcp_server(
    tools: list[Tool],
    cfg: Optional[MCPServeConfig] = None,
    on_call: Optional[OnCall] = None,
) -> Server:
    """Build a low-level MCP :class:`Server` exposing ``tools`` as MCP tools.

    ``tools/list`` maps each Tool's name (verbatim — already sanitized at
    registration), description, and input schema; ``tools/call`` dispatches to
    ``Tool.execute`` and maps the :class:`ToolResult` to a ``CallToolResult``
    (``output`` → text, ``is_error`` propagates). A fresh server is cheap to build,
    so the HTTP path makes one per request (stateless).
    """
    by_name = {t.name: t for t in tools}
    server: Server = Server(
        name=_coalesce(cfg.get("name") if cfg else None, DEFAULT_NAME),
        version=_coalesce(cfg.get("version") if cfg else None, DEFAULT_VERSION),
    )

    @server.list_tools()
    async def _list_tools() -> list[McpToolDef]:
        return [
            McpToolDef(name=t.name, description=t.description, inputSchema=t.input_schema)
            for t in tools
        ]

    # validate_input=False: no input validation — execute() is called directly,
    # matching the JS reference (which does not validate against inputSchema).
    @server.call_tool(validate_input=False)
    async def _call_tool(name: str, arguments: dict[str, Any]) -> CallToolResult:
        tool = by_name.get(name)
        if tool is None:
            # Unknown tool name → the SDK's standard error (InvalidParams / -32602).
            raise McpError(ErrorData(code=INVALID_PARAMS, message=f"Unknown tool: {name}"))
        started = asyncio.get_running_loop().time()
        # A single tool invocation. An execute() throw becomes an is_error result —
        # never a crash.
        try:
            result = await tool.execute(arguments or {})
        except Exception as e:  # noqa: BLE001 — surfaced as an is_error result
            result = ToolResult(output=str(e), is_error=True)
        if on_call is not None:
            try:
                ev = on_call(
                    {
                        "name": tool.name,
                        "source": tool.source,
                        "ms": (asyncio.get_running_loop().time() - started) * 1000.0,
                        "is_error": result.is_error,
                    }
                )
                if inspect.isawaitable(ev):
                    await ev
            except Exception:  # noqa: BLE001 — host callback errors are isolated
                pass
        return CallToolResult(
            content=[TextContent(type="text", text=result.output)],
            isError=result.is_error,
        )

    return server


# ---------------------------------------------------------------------------
# streamable-HTTP — a single stateless request handled through the MCP transport.
# ---------------------------------------------------------------------------


async def _run_stateless(
    server: Server,
    transport: StreamableHTTPServerTransport,
    *,
    task_status: Any = anyio.TASK_STATUS_IGNORED,
) -> None:
    async with transport.connect() as (read_stream, write_stream):
        task_status.started()
        try:
            await server.run(
                read_stream,
                write_stream,
                server.create_initialization_options(),
                stateless=True,
            )
        except Exception:  # noqa: BLE001 — a crashed session never crashes the host
            pass


async def serve_mcp_http_request(
    server: Server,
    method: str,
    path: str,
    req_headers: list[tuple[bytes, bytes]],
    body: bytes,
) -> tuple[int, list[tuple[bytes, bytes]], bytes]:
    """Handle ONE streamable-HTTP MCP request through a fresh stateless transport,
    returning ``(status, headers, body)`` for the caller to write to the socket.

    Mirrors the JS serve.ts ``/mcp`` path: a fresh server+transport per request
    (stateless — no session id). Built on top of the MCP SDK's ASGI transport via
    a minimal in-memory ASGI bridge, so the sync HTTP server in serve.py can drive it.
    """
    captured: dict[str, Any] = {"status": 500, "headers": [], "body": bytearray()}

    async def receive() -> dict[str, Any]:
        return {"type": "http.request", "body": body, "more_body": False}

    async def send(message: dict[str, Any]) -> None:
        if message["type"] == "http.response.start":
            captured["status"] = message["status"]
            captured["headers"] = message.get("headers", [])
        elif message["type"] == "http.response.body":
            captured["body"].extend(message.get("body", b""))

    scope: dict[str, Any] = {
        "type": "http",
        "asgi": {"version": "3.0", "spec_version": "2.3"},
        "http_version": "1.1",
        "method": method,
        "scheme": "http",
        "path": path,
        "raw_path": path.encode("ascii"),
        "query_string": b"",
        "root_path": "",
        "headers": req_headers,
        "server": ("127.0.0.1", 0),
        "client": ("127.0.0.1", 0),
    }

    transport = StreamableHTTPServerTransport(
        mcp_session_id=None, is_json_response_enabled=True
    )
    async with anyio.create_task_group() as tg:
        await tg.start(_run_stateless, server, transport)
        await transport.handle_request(scope, receive, send)
        await transport.terminate()
        tg.cancel_scope.cancel()

    return captured["status"], captured["headers"], bytes(captured["body"])
